modules/mill-codegen-plugin/test/src/smithy4s/codegen/mill/Smithy4sModuleSpec.scala

```diff
@@ -0,0 +1,122 @@
+package smithy4s.codegen.mill
+
+import mill.testkit.MillTestKit
+import mill.scalalib._
+import mill._
+import munit.Location
+import sourcecode.FullName
+
+class Smithy4sModuleSpec extends munit.FunSuite {
+  object testKit extends MillTestKit
+
+  val coreDep =
+    ivy"com.disneystreaming.smithy4s::smithy4s-core:${smithy4s.codegen.BuildInfo.version}"
+
+  object foo extends testKit.BaseModule with Smithy4sModule {
+    override def scalaVersion = "2.13.8"
+    override def ivyDeps = Agg(coreDep)
+  }
+
+  object bar extends testKit.BaseModule with Smithy4sModule {
+    override def moduleDeps = Seq(foo)
+    override def scalaVersion = "2.13.8"
+    override def ivyDeps = Agg(coreDep)
+  }
+
+  test("codegen runs") {
+    val ev = testKit.staticTestEvaluator(foo)(FullName("codegen-runs"))
+    os.write.over(
+      foo.millSourcePath / "smithy" / "foo.smithy",
+      s"""|$$version: "2"
+          |
+          |namespace basic
+          |
+          |string MyNewString""".stripMargin,
+      createFolders = true
+    )
+
+    compileWorks(foo, ev)
+    checkFileExist(
+      ev.outPath / "smithy4sOutputDir.dest" / "scala" / "basic" / "MyNewString.scala",
+      shouldExist = true
+    )
+  }
+
+  test("multi-module codegen works") {
```

Comment:
the downside of this approach is that because you have sort of a virtual project, you have to write source files to disk before running your task

at least that's how I got it to work

Reply:
@daddykotex you can shove the source files in a resource folder and set the modules to point to those, you shouldn't have to programmatically write them.

Reply:
In particular, you can override the `millSourcePath` field on your instance of `BaseModule`, and you can override the `target` field on the `testkit` instance to point to a temp folder or something in `out` or whatnot.

Reply:
that worked, good awareness, did not think of overriding the module src path